src/test/java/org/reactome/addlinks/test/TestEnsemblBioMartRetriever.java

```diff
@@ -0,0 +1,97 @@
+
+import org.junit.Test;
+import org.reactome.addlinks.dataretrieval.ensembl.EnsemblBioMartRetriever;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.*;
+
+import java.io.ByteArrayOutputStream;
+import java.io.PrintStream;
+import java.net.URI;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+public class TestEnsemblBioMartRetriever {
+
+    private static final String testDirectory = "/tmp/test_ensembl_biomart_mapping_service/";
+    private static final String expectedErrorMessage = "java.lang.Exception: BioMart query failed with message: Query ERROR";
+
+    //TODO Move these to IT?
+    @Test
+    public void testEnsemblBioMartRetrieverDownloadsAndStoresData() throws Exception {
+
+        EnsemblBioMartRetriever retriever = new EnsemblBioMartRetriever();
+        System.setProperty("config.location", "src/test/resources/addlinksTest-btau.properties");
+        retriever.setDataURL(new URI("http://www.ensembl.org/biomart/martservice?"));
+        retriever.setFetchDestination(testDirectory);
+        retriever.downloadData();
+
+        Path fetchDestinationMicroarray = Paths.get(testDirectory + "btaurus_microarray");
+        boolean microarrayFileExists = Files.exists(fetchDestinationMicroarray);
+        Files.delete(fetchDestinationMicroarray);
+        Path fetchDestinationUniprot = Paths.get(testDirectory + "btaurus_uniprot");
+        boolean uniprotFileExists = Files.exists(fetchDestinationUniprot);
+        Files.delete(fetchDestinationUniprot);
+
+        assertThat(microarrayFileExists, is(equalTo(true)));
+        assertThat(uniprotFileExists, is(equalTo(true)));
+    }
+
+    /**
+     * This test is for cases where one of (multiple) queries fails due to the data not existing. It should throw the
+     * Exception message but still continue downloading data. The best example, at time of writing (Jan 2020), is
+     * S. cerevisiae for the UniProt download. The Uniprot files are comprised of SwissProt and TrEMBL data, but only
+     * SwissProt data exists for S. cerevisiae. This means an Exception should be thrown during the TrEMBL
+     * attempt but the UniProt file should still exist after successfully downloading SwissProt data.
+     * @throws Exception
+     */
+    @Test
+    public void testEnsemblBioMartRetrieverDownloadsPartiallyAvailableData() throws Exception {
+        EnsemblBioMartRetriever retriever = new EnsemblBioMartRetriever();
+        System.setProperty("config.location", "src/test/resources/addlinksTest-scer.properties");
+        final ByteArrayOutputStream errStream = new ByteArrayOutputStream();
+        System.setErr(new PrintStream(errStream));
+        retriever.setDataURL(new URI("http://www.ensembl.org/biomart/martservice?"));
+        retriever.setFetchDestination(testDirectory);
+        retriever.downloadData();
+
+        boolean exceptionHappened = errStream.toString().contains(expectedErrorMessage);
```

Comment:
For testing exceptions, this is a preferred way with JUnit4:

```
@Rule
public ExpectedException expectedException = ExpectedException.none();

@Test
public void testEnsemblBioMartRetrieverDownloadsPartiallyAvailableData() throws Exception {
    ...
    expectedException.expect(Exception.class);
    expectedException.expectMessage("BioMart query failed with message: Query ERROR");
    [method call that generates the expected exception]
}
```

Reply:
Oh sorry, I think I misunderstood.  Is this test looking for a printed exception message?  Does the error get printed to STDERR but not thrown?

Reply:
Yes its looking for the `BioMart query failed...` message to have been thrown, so I just redirected the STDERR stream and check that the phrase exists. Basically, there needs to still be downloaded data while also outputting an error message

Reply:
Okay, sounds good.